File: js/inventory.js
// =============================================================
// inventory.js — Inklub Store Manager
// Inventario Screen Logic
// =============================================================
// Handles everything on the Inventario screen:
//   - Loading and displaying inventory items
//   - Category filter chips
//   - Adding new categories dynamically
//   - Adding new items (regular units OR vinyl rolls)
//   - Updating item quantities / remaining inches
//   - Updating minimum stock thresholds
//   - Deleting items and categories
//
// VINYL ROLL SUPPORT:
// Items in categories with tracking_type = 'roll' are displayed
// and edited differently from regular unit items:
//   - Display: "85in / 120in" instead of "10 units"
//   - Add form: shows roll-specific fields (color, total, remaining)
//   - Edit prompt: asks for remaining inches instead of quantity
//   - Bar fill: based on remaining/total percentage
//   - Minimum: shown in inches instead of units
//
// The add form detects the category type automatically and
// switches between regular and roll fields dynamically.
// =============================================================

import {
  getInventory, addInventoryItem, updateItemQuantity,
  updateItemThreshold, deleteInventoryItem,
  getCategories, addCategory, deleteCategory
} from './api.js';


// Track the active category filter (null = show all)
let activeCategory = null;

// Store categories so forms can use them without re-fetching
let currentCategories = [];

// Store items for filtering without re-fetching
let currentItems = [];

// Track the current market so cancel/back buttons always
// know which market to return to
let currentMarket = null;


// -------------------------------------------------------------
// initInventory(marketId)
// -------------------------------------------------------------
// Entry point for the Inventario screen.
// Loads categories and items in parallel for speed.
// Called by dashboard.js whenever this screen is shown
// or the market toggle changes.
// -------------------------------------------------------------
export async function initInventory(marketId) {
  activeCategory = null;
  currentMarket = marketId;

  const container = document.getElementById('screen-inventario');
  if (!container) return;

  container.innerHTML = '<div class="loading">Cargando inventario...</div>';

  try {
    // Promise.all loads both at the same time — faster than sequential calls
    [currentCategories, currentItems] = await Promise.all([
      getCategories(),
      getInventory(marketId)
    ]);

    renderInventoryScreen(marketId);
  } catch (error) {
    container.innerHTML = '<div class="empty-state">Error al cargar el inventario.</div>';
    console.error('Inventory error:', error);
  }
}


// -------------------------------------------------------------
// renderInventoryScreen(marketId)
// -------------------------------------------------------------
// Builds the full inventory screen HTML and injects it.
// Called after data loads and after any update.
// Handles both regular items and vinyl roll items in the same list.
// -------------------------------------------------------------
function renderInventoryScreen(marketId) {
  currentMarket = marketId;
  const container = document.getElementById('screen-inventario');

  // Count stock issues — roll items use remaining_inches vs threshold,
  // regular items use quantity vs threshold
  const lowStockItems = currentItems.filter(i => {
    if (i.tracking_type === 'roll') {
      return i.remaining_inches > 0 && i.remaining_inches <= i.threshold;
    }
    return i.quantity > 0 && i.quantity <= i.threshold;
  });

  const outOfStockItems = currentItems.filter(i => {
    if (i.tracking_type === 'roll') return (i.remaining_inches || 0) === 0;
    return i.quantity === 0;
  });

  // Build the alert banner — only shown if there are stock issues
  const alertHTML = (lowStockItems.length > 0 || outOfStockItems.length > 0) ? `
    <div class="alert-banner">
      <div class="alert-label">Alerta de stock</div>
      <div class="alert-text">
        ${outOfStockItems.length > 0 ? `${outOfStockItems.length} agotado(s). ` : ''}
        ${lowStockItems.length > 0 ? `${lowStockItems.length} con stock bajo.` : ''}
      </div>
    </div>
  ` : '';

  // Filter items by the active category chip
  const filteredItems = activeCategory
    ? currentItems.filter(i => i.category_id === activeCategory)
    : currentItems;

  // Build category filter chips — one per category
  const allChipActive = activeCategory === null ? 'active' : '';
  const categoryChipsHTML = currentCategories.map(cat => `
    <button class="btn btn-filter ${activeCategory === cat.id ? 'active' : ''}"
            onclick="window._invSetCategory(${cat.id}, '${marketId}')">
      ${cat.label}
    </button>
  `).join('');

  const itemsHTML = filteredItems.length > 0
    ? filteredItems.map(item => renderInventoryCard(item, marketId)).join('')
    : '<div class="empty-state">No hay productos en esta categoría.</div>';

  const activeCatLabel = activeCategory
    ? currentCategories.find(c => c.id === activeCategory)?.label || 'Productos'
    : 'Todos los productos';

  container.innerHTML = `
    ${alertHTML}

    <div class="toolbar">
      <button class="btn btn-filter ${allChipActive}"
              onclick="window._invSetCategory(null, '${marketId}')">Todo</button>
      ${categoryChipsHTML}
      <button class="btn btn-dashed"
              onclick="window._invShowAddCategory('${marketId}')">+ Categoría</button>
      <button class="btn btn-secondary ml-auto"
              onclick="window._invShowAddItem('${marketId}')">+ Agregar</button>
    </div>

    <div class="section-title mb-sm">${activeCatLabel} — ${filteredItems.length} productos</div>

    ${itemsHTML}

    <div class="footer-brand">Elevate your style.</div>
  `;

  // Expose all action handlers on window so onclick attributes
  // in the injected HTML can call them (module functions aren't global)
  window._invSetCategory = (catId, market) => { activeCategory = catId; renderInventoryScreen(market); };
  window._invShowAddCategory = (market) => showAddCategoryForm(market);
  window._invShowAddItem = (market) => showAddItemForm(market);
  window._invUpdateQty = (itemId, market) => showUpdateStockForm(itemId, market);
  window._invUpdateMin = (itemId, market) => showUpdateThresholdForm(itemId, market);
  window._invDeleteItem = (itemId, market) => handleDeleteItem(itemId, market);
  // Cancel button handler — returns to inventory list
  window._invCancel = (market) => initInventory(market);
}


// -------------------------------------------------------------
// renderInventoryCard(item, marketId)
// -------------------------------------------------------------
// Returns the HTML for a single inventory card.
// Renders differently based on tracking_type:
//
// 'roll' items show:
//   - "85in / 120in" quantity display
//   - Bar fill based on remaining/total percentage
//   - Roll color badge
//   - "Actualizar rollo" button label
//   - Minimum shown in inches
//
// 'units' items show:
//   - "10 units" quantity display
//   - Bar fill based on qty vs threshold*3
//   - "Editar stock" button label
//   - Minimum shown in units
// -------------------------------------------------------------
function renderInventoryCard(item, marketId) {
  const isRoll = item.tracking_type === 'roll';

  // --- Determine stock status ---
  let status;
  if (isRoll) {
    const remaining = item.remaining_inches || 0;
    if (remaining === 0) status = 'out';
    else if (remaining <= item.threshold) status = 'low';
    else status = 'ok';
  } else {
    if (item.quantity === 0) status = 'out';
    else if (item.quantity <= item.threshold) status = 'low';
    else status = 'ok';
  }

  // --- Calculate progress bar fill percentage ---
  let fillPct;
  if (isRoll) {
    // For rolls: remaining / total * 100
    const total = item.total_inches || 0;
    const remaining = item.remaining_inches || 0;
    fillPct = total > 0 ? Math.min(Math.round((remaining / total) * 100), 100) : 0;
  } else {
    // For regular items: qty / (threshold * 3) as a reference max
    const maxRef = Math.max(item.threshold * 3, item.quantity);
    fillPct = maxRef > 0 ? Math.min(Math.round((item.quantity / maxRef) * 100), 100) : 0;
  }

  // --- Build quantity display label ---
  const qtyDisplay = isRoll
    ? `${item.remaining_inches || 0}in / ${item.total_inches || 0}in`
    : `${item.quantity} ${item.unit}`;

  // --- Build minimum display label ---
  const minDisplay = isRoll
    ? `${item.threshold} in`
    : `${item.threshold} ${item.unit}`;

  // --- Card CSS class based on stock status ---
  const cardClass = status === 'out'
    ? 'card out-of-stock'
    : status === 'low'
      ? 'card warning'
      : 'card';

  // --- Roll color badge (only for vinyl roll items) ---
  const colorLabel = isRoll && item.roll_color
    ? `<span class="inv-category" style="margin-left:4px;">${item.roll_color}</span>`
    : '';

  return `
    <div class="${cardClass}" id="inv-${item.id}">
      <div class="inv-card-top">
        <div class="inv-dot ${status}"></div>
        <span class="inv-name">${item.name}</span>
        <span class="inv-category">${item.category_label}</span>
        ${colorLabel}
      </div>

      <div class="inv-bar-row">
        <div class="inv-bar-track">
          <div class="inv-bar-fill ${status}" style="width: ${fillPct}%"></div>
        </div>
        <span class="inv-qty">${qtyDisplay}</span>
      </div>

      <div class="inv-minimum-row">
        <span class="inv-minimum-label">Mínimo:</span>
        <span class="inv-minimum-value">${minDisplay}</span>
      </div>

      <div class="card-actions">
        <button class="btn btn-primary"
                onclick="window._invUpdateQty(${item.id}, '${marketId}')">
          ${isRoll ? 'Actualizar rollo' : 'Editar stock'}
        </button>
        <button class="btn btn-secondary"
                onclick="window._invUpdateMin(${item.id}, '${marketId}')">
          Mínimo
        </button>
        <button class="btn btn-danger"
                onclick="window._invDeleteItem(${item.id}, '${marketId}')">
          Eliminar
        </button>
      </div>
    </div>
  `;
}


// -------------------------------------------------------------
// showUpdateStockForm(itemId, marketId)
// -------------------------------------------------------------
// Shows a prompt to update stock.
// For roll items: asks for remaining inches
// For regular items: asks for new quantity
// Uses native browser prompt() for simplicity — works on mobile
// without needing a custom modal.
// -------------------------------------------------------------
async function showUpdateStockForm(itemId, marketId) {
  const item = currentItems.find(i => i.id === itemId);
  if (!item) return;

  const isRoll = item.tracking_type === 'roll';
  let newValue;

  if (isRoll) {
    // For rolls, ask how many inches remain after the latest use
    newValue = prompt(
      `Actualizar rollo "${item.name}"\n` +
      `Total del rollo: ${item.total_inches}in\n` +
      `Restante actual: ${item.remaining_inches}in\n\n` +
      `¿Cuántas pulgadas quedan ahora?`,
      item.remaining_inches
    );
  } else {
    newValue = prompt(
      `Actualizar stock de "${item.name}"\n` +
      `Cantidad actual: ${item.quantity} ${item.unit}\n\n` +
      `Nueva cantidad:`,
      item.quantity
    );
  }

  // User cancelled — null means they pressed Cancel
  if (newValue === null || newValue === '') return;

  const parsed = parseFloat(newValue);
  if (isNaN(parsed) || parsed < 0) {
    alert('Por favor ingresa un número válido (0 o más).');
    return;
  }

  try {
    if (isRoll) {
      // Pass remaining_inches for roll items
      await updateItemQuantity(itemId, null, parsed);
    } else {
      // Pass quantity for regular items
      await updateItemQuantity(itemId, parsed);
    }
    // Reload to reflect the updated values
    await initInventory(marketId);
  } catch (error) {
    alert('Error al actualizar. Inténtalo de nuevo.');
    console.error('Update stock error:', error);
  }
}


// -------------------------------------------------------------
// showUpdateThresholdForm(itemId, marketId)
// -------------------------------------------------------------
// Shows a prompt to update the minimum stock threshold.
// For roll items the threshold is in inches.
// For regular items it's in units.
// An email alert is sent when stock drops to or below this value.
// -------------------------------------------------------------
async function showUpdateThresholdForm(itemId, marketId) {
  const item = currentItems.find(i => i.id === itemId);
  if (!item) return;

  const isRoll = item.tracking_type === 'roll';
  const unit = isRoll ? 'pulgadas' : item.unit;

  const newThreshold = prompt(
    `Mínimo de "${item.name}"\n` +
    `Mínimo actual: ${item.threshold} ${unit}\n\n` +
    `Nuevo mínimo (recibirás una alerta cuando llegues a este número):`,
    item.threshold
  );

  if (newThreshold === null || newThreshold === '') return;

  const threshold = parseFloat(newThreshold);
  if (isNaN(threshold) || threshold < 0) {
    alert('Por favor ingresa un número válido.');
    return;
  }

  try {
    await updateItemThreshold(itemId, threshold);
    await initInventory(marketId);
  } catch (error) {
    alert('Error al actualizar el mínimo.');
    console.error('Update threshold error:', error);
  }
}


// -------------------------------------------------------------
// showAddItemForm(marketId)
// -------------------------------------------------------------
// Replaces the inventory screen with a form to add a new item.
//
// The form dynamically shows different fields based on the
// selected category's tracking_type:
//   - 'units' → shows quantity + threshold fields
//   - 'roll'  → shows color, total inches, remaining, threshold
//
// The switch happens in real time via the onchange handler on
// the category select — window._invCategoryChanged().
// -------------------------------------------------------------
function showAddItemForm(marketId) {
  const container = document.getElementById('screen-inventario');

  // Build category options — embed tracking_type as data attribute
  // so the JS can read it when the user selects a category
  const categoryOptionsHTML = currentCategories.map(cat =>
    `<option value="${cat.id}" data-type="${cat.tracking_type}">${cat.label}</option>`
  ).join('');

  container.innerHTML = `
    <div class="form-card">
      <div class="form-title">Nuevo producto</div>

      <label class="field-label">Categoría</label>
      <select class="field-input" id="new-item-category"
              onchange="window._invCategoryChanged()">
        ${categoryOptionsHTML}
      </select>

      <label class="field-label">Nombre</label>
      <input class="field-input" id="new-item-name"
             placeholder="Ej. T-shirt Blanca, Vinilo Negro Rollo 1" />

      <!-- Regular unit fields — shown for non-roll categories -->
      <div id="regular-fields">
        <div class="field-row">
          <div>
            <label class="field-label">Cantidad inicial</label>
            <input class="field-input" id="new-item-qty"
                   type="number" min="0" value="0" />
          </div>
          <div>
            <label class="field-label">Mínimo (unidades)</label>
            <input class="field-input" id="new-item-threshold"
                   type="number" min="0" value="5" />
          </div>
        </div>
      </div>

      <!-- Vinyl roll fields — shown when a roll category is selected -->
      <div id="roll-fields" style="display:none;">
        <label class="field-label">Color del rollo</label>
        <input class="field-input" id="new-roll-color"
               placeholder="Ej. Negro, Rojo, Blanco, Dorado" />

        <div class="field-row">
          <div>
            <label class="field-label">Largo total (pulgadas)</label>
            <input class="field-input" id="new-roll-total"
                   type="number" min="0" step="0.5"
                   placeholder="Ej. 120" />
          </div>
          <div>
            <label class="field-label">Restante actual (in)</label>
            <input class="field-input" id="new-roll-remaining"
                   type="number" min="0" step="0.5"
                   placeholder="Igual al total si es nuevo" />
          </div>
        </div>

        <div class="field-row">
          <div>
            <label class="field-label">Mínimo (pulgadas)</label>
            <input class="field-input" id="new-roll-threshold"
                   type="number" min="0" value="12" />
          </div>
        </div>
      </div>

    </div>

    <div class="flex gap-sm mt-md">
      <button class="btn btn-primary flex-1"
              onclick="window._invSaveNewItem('${marketId}')">
        Guardar producto
      </button>
      <button class="btn btn-secondary"
              onclick="window._invCancel('${marketId}')">
        Cancelar
      </button>
    </div>
  `;

  // -------------------------------------------------------------
  // _invCategoryChanged — toggles roll vs regular fields
  // -------------------------------------------------------------
  // Reads the data-type attribute of the selected option and
  // shows/hides the appropriate field groups.
  // -------------------------------------------------------------
  window._invCategoryChanged = () => {
    const select = document.getElementById('new-item-category');
    const selected = select.options[select.selectedIndex];
    const isRoll = selected.dataset.type === 'roll';

    document.getElementById('regular-fields').style.display = isRoll ? 'none' : 'block';
    document.getElementById('roll-fields').style.display = isRoll ? 'block' : 'none';
  };

  // Trigger on load to set the correct initial state
  window._invCategoryChanged();

  // -------------------------------------------------------------
  // _invSaveNewItem — reads the form and calls the API
  // -------------------------------------------------------------
  window._invSaveNewItem = async (market) => {
    const select = document.getElementById('new-item-category');
    const selected = select.options[select.selectedIndex];
    const categoryId = parseInt(select.value);
    const isRoll = selected.dataset.type === 'roll';
    const name = document.getElementById('new-item-name').value.trim();

    if (!name) {
      alert('El nombre del producto es obligatorio.');
      return;
    }

    try {
      if (isRoll) {
        // --- Vinyl roll save ---
        const totalInches = parseFloat(document.getElementById('new-roll-total').value) || 0;
        const remainingInches = parseFloat(document.getElementById('new-roll-remaining').value) || totalInches;
        const threshold = parseFloat(document.getElementById('new-roll-threshold').value) || 12;
        const rollColor = document.getElementById('new-roll-color').value.trim();

        if (totalInches === 0) {
          alert('El largo total del rollo es obligatorio.');
          return;
        }

        await addInventoryItem({
          market_id: market,
          category_id: categoryId,
          name,
          roll_color: rollColor,
          total_inches: totalInches,
          remaining_inches: remainingInches,
          threshold
        });

      } else {
        // --- Regular item save ---
        const quantity = parseInt(document.getElementById('new-item-qty').value) || 0;
        const threshold = parseInt(document.getElementById('new-item-threshold').value) || 5;

        await addInventoryItem({
          market_id: market,
          category_id: categoryId,
          name,
          quantity,
          threshold
        });
      }

      // Return to inventory list after saving
      await initInventory(market);

    } catch (error) {
      alert('Error al guardar el producto.');
      console.error('Add item error:', error);
    }
  };
}


// -------------------------------------------------------------
// showAddCategoryForm(marketId)
// -------------------------------------------------------------
// Uses browser prompts to collect the category name and type.
// Asks if it's a vinyl roll category — if yes, sets tracking_type
// to 'roll' and unit to 'in'. Otherwise uses units.
// -------------------------------------------------------------
async function showAddCategoryForm(marketId) {
  const label = prompt('Nombre de la nueva categoría:\n(Ej. Gorras, Ornamentos, Tote bags...)');
  if (!label || !label.trim()) return;

  // Ask if this is a vinyl roll category
  // confirm() returns true for OK, false for Cancel
  const isRoll = confirm(
    `¿"${label.trim()}" es una categoría de rollos de vinilo?\n\n` +
    `OK = Sí, se trackea por pulgadas\n` +
    `Cancelar = No, se trackea por unidades`
  );

  let unit;
  if (!isRoll) {
    unit = prompt(
      `Unidad para "${label.trim()}":\n(Ej. units, pcs, pairs)`,
      'units'
    );
    if (!unit) return;
  }

  try {
    await addCategory({
      label: label.trim(),
      unit: isRoll ? 'in' : unit.trim(),
      tracking_type: isRoll ? 'roll' : 'units'
    });
    // Reload so the new category chip appears immediately
    await initInventory(marketId);
  } catch (error) {
    if (error.message.includes('already exists')) {
      alert('Ya existe una categoría con ese nombre.');
    } else {
      alert('Error al crear la categoría.');
    }
    console.error('Add category error:', error);
  }
}


// -------------------------------------------------------------
// handleDeleteItem(itemId, marketId)
// -------------------------------------------------------------
// Asks for confirmation before permanently deleting an item.
// -------------------------------------------------------------
async function handleDeleteItem(itemId, marketId) {
  const item = currentItems.find(i => i.id === itemId);
  const name = item ? item.name : 'este producto';

  if (!confirm(`¿Segura que quieres eliminar "${name}"?`)) return;

  try {
    await deleteInventoryItem(itemId);
    await initInventory(marketId);
  } catch (error) {
    alert('Error al eliminar el producto.');
    console.error('Delete item error:', error);
  }
}

File: backend/routes/inventory.py
# =============================================================
# routes/inventory.py — Inventory Management Routes
# =============================================================
# Handles all inventory operations:
#   - Listing items (with optional category filter)
#   - Adding new items (regular units OR vinyl rolls)
#   - Updating quantities / remaining inches
#   - Updating minimum stock thresholds
#   - Managing categories (add / delete)
#   - Triggering low-stock notifications
#
# VINYL ROLL TRACKING:
# Items in categories with tracking_type = 'roll' are measured
# in inches rather than whole units. They have three extra fields:
#   - total_inches     : full length of the roll when new
#   - remaining_inches : how many inches are left
#   - roll_color       : e.g. "Negro", "Rojo", "Blanco"
#
# When a Trello card moves to Entregados, inventory_deduction.py
# calls this module's update logic to deduct the used inches.
# =============================================================

from flask import Blueprint, jsonify, request
from db import get_db
from routes.auth import require_auth
from notifications import send_low_stock_alert

inventory_bp = Blueprint("inventory", __name__)


# -------------------------------------------------------------
# GET /api/inventory/<market_id>
# -------------------------------------------------------------
# Returns all inventory items for a market, joined with their
# category info (label, unit, tracking_type).
# Optional query param: ?category_id=3 to filter by category.
#
# tracking_type is included so the frontend knows whether to
# display the item as units or as remaining/total inches.
# -------------------------------------------------------------
@inventory_bp.route("/<market_id>", methods=["GET"])
@require_auth
def get_inventory(market_id):
    category_id = request.args.get("category_id")

    with get_db() as conn:
        if category_id:
            # Filter by category if the frontend sent one
            items = conn.execute("""
                SELECT i.*, c.label AS category_label, c.unit, c.tracking_type
                FROM inventory i
                JOIN categories c ON c.id = i.category_id
                WHERE i.market_id = %s AND i.category_id = %s
                ORDER BY c.label, i.name
            """, (market_id, category_id)).fetchall()
        else:
            # Return all items for this market
            items = conn.execute("""
                SELECT i.*, c.label AS category_label, c.unit, c.tracking_type
                FROM inventory i
                JOIN categories c ON c.id = i.category_id
                WHERE i.market_id = %s
                ORDER BY c.label, i.name
            """, (market_id,)).fetchall()

    return jsonify(list(items)), 200


# -------------------------------------------------------------
# POST /api/inventory
# -------------------------------------------------------------
# Adds a new inventory item. Behavior differs based on whether
# the category is a regular item or a vinyl roll.
#
# Expected JSON body for a REGULAR item:
# {
#   "market_id":   "ca",
#   "category_id": 1,
#   "name":        "T-shirt Blanca",
#   "quantity":    10,
#   "threshold":   3
# }
#
# Expected JSON body for a VINYL ROLL:
# {
#   "market_id":         "ca",
#   "category_id":       5,
#   "name":              "Vinilo Negro Rollo 1",
#   "roll_color":        "Negro",
#   "total_inches":      120,
#   "remaining_inches":  120,
#   "threshold":         12
# }
#
# For roll items, quantity is stored as a percentage (0-100)
# derived from remaining/total — this is used for the bar
# display in the frontend. The actual remaining inches are
# stored separately in remaining_inches.
# -------------------------------------------------------------
@inventory_bp.route("", methods=["POST"])
@require_auth
def add_item():
    data = request.get_json()

    # Name, market, and category are always required
    required = ["market_id", "category_id", "name"]
    missing = [f for f in required if data.get(f) is None]
    if missing:
        return jsonify({"error": f"Campos requeridos: {', '.join(missing)}"}), 400

    with get_db() as conn:
        # Look up the category to determine tracking type
        category = conn.execute("""
            SELECT tracking_type FROM categories WHERE id = %s
        """, (data["category_id"],)).fetchone()

        if not category:
            return jsonify({"error": "Categoría no encontrada."}), 404

        is_roll = category["tracking_type"] == "roll"

        if is_roll:
            # --- Vinyl roll item ---
            total_inches     = data.get("total_inches", 0)
            remaining_inches = data.get("remaining_inches", total_inches)
            roll_color       = data.get("roll_color", "")

            # quantity stores the fill percentage (0-100) for the progress bar
            # e.g. if 90in remain out of 120in total → quantity = 75
            quantity = round((remaining_inches / total_inches * 100)) if total_inches > 0 else 0

            item = conn.execute("""
                INSERT INTO inventory (
                    market_id, category_id, name, quantity, threshold,
                    total_inches, remaining_inches, roll_color
                ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s)
                RETURNING *
            """, (
                data["market_id"],
                data["category_id"],
                data["name"],
                quantity,
                data.get("threshold", 12),  # Default threshold: 12 inches for vinyl
                total_inches,
                remaining_inches,
                roll_color
            )).fetchone()

        else:
            # --- Regular unit item ---
            item = conn.execute("""
                INSERT INTO inventory (
                    market_id, category_id, name, quantity, threshold
                ) VALUES (%s, %s, %s, %s, %s)
                RETURNING *
            """, (
                data["market_id"],
                data["category_id"],
                data["name"],
                data.get("quantity", 0),
                data.get("threshold", 5)   # Default threshold: 5 units
            )).fetchone()

    return jsonify({"message": "Producto agregado.", "item": dict(item)}), 201


# -------------------------------------------------------------
# PATCH /api/inventory/<id>/quantity
# -------------------------------------------------------------
# Updates stock for an item. Behavior differs by tracking type:
#
# For REGULAR items:
#   Expects: { "quantity": 10 }
#   Updates the quantity column directly.
#
# For VINYL ROLLS:
#   Expects: { "remaining_inches": 85.5 }
#   Updates remaining_inches and recalculates quantity (%).
#
# After updating, checks if the item is now at or below its
# threshold and triggers a low-stock email alert if so.
# -------------------------------------------------------------
@inventory_bp.route("/<int:item_id>/quantity", methods=["PATCH"])
@require_auth
def update_quantity(item_id):
    data = request.get_json()

    with get_db() as conn:
        # Fetch the item with its category info to determine tracking type
        item = conn.execute("""
            SELECT i.*, c.tracking_type, c.unit
            FROM inventory i
            JOIN categories c ON c.id = i.category_id
            WHERE i.id = %s
        """, (item_id,)).fetchone()

        if not item:
            return jsonify({"error": "Producto no encontrado."}), 404

        if item["tracking_type"] == "roll":
            # --- Vinyl roll update ---
            if "remaining_inches" not in data:
                return jsonify({
                    "error": "remaining_inches es requerido para rollos de vinilo."
                }), 400

            remaining = float(data["remaining_inches"])
            total     = item["total_inches"] or 0

            # Recalculate the percentage for the progress bar
            quantity = round((remaining / total * 100)) if total > 0 else 0

            updated = conn.execute("""
                UPDATE inventory
                SET remaining_inches = %s, quantity = %s
                WHERE id = %s
                RETURNING *
            """, (remaining, quantity, item_id)).fetchone()

            # Check if remaining inches dropped below threshold
            if remaining <= item["threshold"]:
                setting = conn.execute(
                    "SELECT value FROM settings WHERE key = 'notification_email'"
                ).fetchone()
                if setting:
                    send_low_stock_alert(
                        to_email=setting["value"],
                        item_name=item["name"],
                        quantity=remaining,
                        unit="pulgadas",
                        market_id=item["market_id"],
                        threshold=item["threshold"]
                    )

        else:
            # --- Regular item update ---
            if "quantity" not in data or data["quantity"] < 0:
                return jsonify({"error": "Cantidad inválida."}), 400

            updated = conn.execute("""
                UPDATE inventory SET quantity = %s WHERE id = %s RETURNING *
            """, (data["quantity"], item_id)).fetchone()

            # Check if quantity dropped below threshold
            if updated["quantity"] <= item["threshold"]:
                setting = conn.execute(
                    "SELECT value FROM settings WHERE key = 'notification_email'"
                ).fetchone()
                if setting:
                    send_low_stock_alert(
                        to_email=setting["value"],
                        item_name=item["name"],
                        quantity=updated["quantity"],
                        unit=item["unit"],
                        market_id=item["market_id"],
                        threshold=item["threshold"]
                    )

    return jsonify({"message": "Stock actualizado.", "item": dict(updated)}), 200


# -------------------------------------------------------------
# PATCH /api/inventory/<id>/threshold
# -------------------------------------------------------------
# Updates the minimum stock threshold for an item.
# For vinyl rolls this value is in inches.
# For regular items this value is in units.
# An alert email is sent when stock drops to or below this number.
# -------------------------------------------------------------
@inventory_bp.route("/<int:item_id>/threshold", methods=["PATCH"])
@require_auth
def update_threshold(item_id):
    data = request.get_json()

    if "threshold" not in data or data["threshold"] < 0:
        return jsonify({"error": "Umbral inválido."}), 400

    with get_db() as conn:
        item = conn.execute("""
            UPDATE inventory SET threshold = %s
            WHERE id = %s RETURNING *
        """, (data["threshold"], item_id)).fetchone()

        if not item:
            return jsonify({"error": "Producto no encontrado."}), 404

    return jsonify({"message": "Mínimo actualizado.", "item": dict(item)}), 200


# -------------------------------------------------------------
# DELETE /api/inventory/<id>
# -------------------------------------------------------------
# Permanently deletes an inventory item.
# The frontend asks for confirmation before calling this.
# -------------------------------------------------------------
@inventory_bp.route("/<int:item_id>", methods=["DELETE"])
@require_auth
def delete_item(item_id):
    with get_db() as conn:
        deleted = conn.execute(
            "DELETE FROM inventory WHERE id = %s RETURNING id", (item_id,)
        ).fetchone()

        if not deleted:
            return jsonify({"error": "Producto no encontrado."}), 404

    return jsonify({"message": "Producto eliminado."}), 200


# =============================================================
# CATEGORY MANAGEMENT
# =============================================================

# -------------------------------------------------------------
# GET /api/inventory/categories
# -------------------------------------------------------------
# Returns all categories including tracking_type so the frontend
# knows whether to show roll or unit fields in the add item form.
# -------------------------------------------------------------
@inventory_bp.route("/categories", methods=["GET"])
@require_auth
def get_categories():
    with get_db() as conn:
        categories = conn.execute(
            "SELECT * FROM categories ORDER BY label"
        ).fetchall()
    return jsonify(list(categories)), 200


# -------------------------------------------------------------
# POST /api/inventory/categories
# -------------------------------------------------------------
# Creates a new category. She does this from the UI — no code needed.
#
# Expected JSON body:
# {
#   "label":         "Gorras",
#   "unit":          "units",
#   "tracking_type": "units"   // or "roll" for vinyl
# }
# -------------------------------------------------------------
@inventory_bp.route("/categories", methods=["POST"])
@require_auth
def add_category():
    data = request.get_json()

    if not data.get("label"):
        return jsonify({"error": "El nombre de la categoría es obligatorio."}), 400

    with get_db() as conn:
        # Check for duplicates (case-insensitive)
        existing = conn.execute(
            "SELECT id FROM categories WHERE label ILIKE %s", (data["label"],)
        ).fetchone()

        if existing:
            return jsonify({"error": "Ya existe una categoría con ese nombre."}), 409

        category = conn.execute("""
            INSERT INTO categories (label, unit, tracking_type)
            VALUES (%s, %s, %s)
            RETURNING *
        """, (
            data["label"],
            data.get("unit", "units"),
            data.get("tracking_type", "units")  # Default to units unless specified
        )).fetchone()

    return jsonify({"message": "Categoría agregada.", "category": dict(category)}), 201


# -------------------------------------------------------------
# DELETE /api/inventory/categories/<id>
# -------------------------------------------------------------
# Deletes a category. Will fail if any inventory items still
# reference this category — the frontend shows a helpful error
# message in that case telling her to remove the items first.
# -------------------------------------------------------------
@inventory_bp.route("/categories/<int:category_id>", methods=["DELETE"])
@require_auth
def delete_category(category_id):
    with get_db() as conn:
        # Check if any items still use this category
        in_use = conn.execute(
            "SELECT COUNT(*) AS total FROM inventory WHERE category_id = %s",
            (category_id,)
        ).fetchone()

        if in_use["total"] > 0:
            return jsonify({
                "error": f"No puedes eliminar — {in_use['total']} producto(s) usan esta categoría. "
                         f"Elimínalos o cámbiales la categoría primero."
            }), 409

        deleted = conn.execute(
            "DELETE FROM categories WHERE id = %s RETURNING id", (category_id,)
        ).fetchone()

        if not deleted:
            return jsonify({"error": "Categoría no encontrada."}), 404

    return jsonify({"message": "Categoría eliminada."}), 200